File: 79.word-search.js
/*
 * @lc app=leetcode id=79 lang=javascript
 *
 * [79] Word Search
 */

// @lc code=start
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function (board, word) {
  const m = board.length;
  const n = board[0].length;
  const visited = Array.from({ length: m }, () => []);

  const dfs = (start, i, j) => {
    if (start === word.length) return true;
    if (i < 0 || j < 0 || i >= m || j >= n || visited[i][j]) return false;

    visited[i][j] = true;
    const res =
      board[i][j] === word[start] &&
      (dfs(start + 1, i - 1, j) ||
        dfs(start + 1, i + 1, j) ||
        dfs(start + 1, i, j - 1) ||
        dfs(start + 1, i, j + 1));
    visited[i][j] = false;
    return res;
  };

  for (let i = 0; i < m; ++i) {
    for (let j = 0; j < n; ++j) {
      if (dfs(0, i, j)) return true;
    }
  }
  return false;
};
// @lc code=end

File: solutions/79.word-search.rb
# @param {Character[][]} board
# @param {String} word
# @return {Boolean}
def exist(board, word)
  m = board.length
  n = board[0].length
  visited = Set.new
  word_index = 0

  dfs = lambda do |x, y|
    return true if word_index == word.length
    return false if x<0 || y<0 || x==m || y==n || visited.include?([x,y]) || board[x][y] != word[word_index]

    visited << [x,y]
    word_index += 1
    return true if [[x-1,y], [x+1,y], [x,y-1], [x,y+1]].map{|next_xy| dfs.call *next_xy }.any?
    visited.delete [x,y]
    word_index -= 1
    false
  end

  (0...m).each do |x|
    (0...n).each do |y|
      return true if dfs.call x,y
    end
  end

  return false
end
